File: ED Services/PluginVault/OAuthBase.cs
using System;
using System.Security.Cryptography;
using System.Collections.Generic;
using System.Text;
using System.Web;


using System.Configuration;
using System.Xml;


using EDServices.DataAccessService;
using Microsoft.Xrm.Sdk;

namespace EDServices
{
    public class OAuthBase
    {

        private static string NonceServer = ConfigurationManager.AppSettings["NonceServer"];
       
        public enum SignatureTypes
        {
            HMACSHA1,
            HMACSHA256,
            PLAINTEXT,
            RSASHA1
        }
       
        protected class QueryParameter
        {
            private string name = null;
            private string value = null;
            public QueryParameter(string name, string value)
            {
                this.name = name;
                this.value = value;
            }
            public string Name
            {
                get { return name; }
            }
            public string Value
            {
                get { return value; }
            }
        }
       
        protected class QueryParameterComparer : IComparer<QueryParameter>
        {
            #region IComparer<QueryParameter> Members
            public int Compare(QueryParameter x, QueryParameter y)
            {
                if (x.Name == y.Name)
                {
                    return string.Compare(x.Value, y.Value);
                }
                else
                {
                    return string.Compare(x.Name, y.Name);
                }
            }
            #endregion
        }
        protected const string OAuthVersion = "1.0";
        protected const string OAuthParameterPrefix = "oauth_";
     
        protected const string OAuthConsumerKeyKey = "oauth_consumer_key";
        protected const string OAuthCallbackKey = "oauth_callback";
        protected const string OAuthVersionKey = "oauth_version";
        protected const string OAuthSignatureMethodKey = "oauth_signature_method";
        protected const string OAuthSignatureKey = "oauth_signature";
        protected const string OAuthTimestampKey = "oauth_timestamp";
        protected const string OAuthNonceKey = "oauth_nonce";
        protected const string OAuthTokenKey = "oauth_token";
        protected const string OAuthTokenSecretKey = "oauth_token_secret";
        protected const string HMACSHA1SignatureType = "HMAC-SHA1";
        protected const string HMACSHA256SignatureType = "HMAC-SHA256";
        protected const string PlainTextSignatureType = "PLAINTEXT";
        protected const string RSASHA1SignatureType = "RSA-SHA1";
        protected Random random = new Random();
        protected string unreservedChars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_.~";
        
        private string ComputeHash(HashAlgorithm hashAlgorithm, string data)
        {
            if (hashAlgorithm == null)
            {
                throw new ArgumentNullException("hashAlgorithm");
            }
            if (string.IsNullOrEmpty(data))
            {
                throw new ArgumentNullException("data");
            }
            byte[] dataBuffer = System.Text.Encoding.ASCII.GetBytes(data);
            byte[] hashBytes = hashAlgorithm.ComputeHash(dataBuffer);
            return Convert.ToBase64String(hashBytes);
        }
       
        private List<QueryParameter> GetQueryParameters(string parameters)
        {
            if (parameters.StartsWith("?"))
            {
                parameters = parameters.Remove(0, 1);
            }
            List<QueryParameter> result = new List<QueryParameter>();
            if (!string.IsNullOrEmpty(parameters))
            {
                string[] p = parameters.Split('&');
                foreach (string s in p)
                {
                    if (!string.IsNullOrEmpty(s) && !s.StartsWith(OAuthParameterPrefix))
                    {
                        if (s.IndexOf('=') > -1)
                        {
                            string[] temp = s.Split('=');
                            result.Add(new QueryParameter(temp[0], temp[1]));
                        }
                        else
                        {
                            result.Add(new QueryParameter(s, string.Empty));
                        }
                    }
                }
            }
            return result;
        }
       
        protected string UrlEncode(string value)
        {
            StringBuilder result = new StringBuilder();
            foreach (char symbol in value)
            {
                if (unreservedChars.IndexOf(symbol) != -1)
                {
                    result.Append(symbol);
                }
                else
                {
                    result.Append('%' + String.Format("{0:X2}", (int)symbol));
                }
            }
            return result.ToString();
        }
        
        protected string NormalizeRequestParameters(IList<QueryParameter> parameters)
        {
            StringBuilder sb = new StringBuilder();
            QueryParameter p = null;
            for (int i = 0; i < parameters.Count; i++)
            {
                p = parameters[i];
                sb.AppendFormat("{0}={1}", p.Name, p.Value);
                if (i < parameters.Count - 1)
                {
                    sb.Append("&");
                }
            }
            return sb.ToString();
        }
        
        public string GenerateSignatureBase(Uri url, string consumerKey, string token, string tokenSecret, string httpMethod, string timeStamp, string nonce, string signatureType, out string normalizedUrl, out string normalizedRequestParameters)
        {
            if (token == null)
            {
                token = string.Empty;
            }
            if (tokenSecret == null)
            {
                tokenSecret = string.Empty;
            }
            if (string.IsNullOrEmpty(consumerKey))
            {
                throw new ArgumentNullException("consumerKey");
            }
            if (string.IsNullOrEmpty(httpMethod))
            {
                throw new ArgumentNullException("httpMethod");
            }
            if (string.IsNullOrEmpty(signatureType))
            {
                throw new ArgumentNullException("signatureType");
            }
            normalizedUrl = null;
            normalizedRequestParameters = null;
            List<QueryParameter> parameters = GetQueryParameters(url.Query);
            parameters.Add(new QueryParameter(OAuthVersionKey, OAuthVersion));
            parameters.Add(new QueryParameter(OAuthNonceKey, nonce));
            parameters.Add(new QueryParameter(OAuthTimestampKey, timeStamp));
            parameters.Add(new QueryParameter(OAuthSignatureMethodKey, signatureType));
            parameters.Add(new QueryParameter(OAuthConsumerKeyKey, consumerKey));
            if (!string.IsNullOrEmpty(token))
            {
                parameters.Add(new QueryParameter(OAuthTokenKey, token));
            }
            parameters.Sort(new QueryParameterComparer());
            normalizedUrl = string.Format("{0}://{1}", url.Scheme, url.Host);
            if (!((url.Scheme == "http" && url.Port == 80) || (url.Scheme == "https" && url.Port == 443)))
            {
                normalizedUrl += ":" + url.Port;
            }
            normalizedUrl += url.AbsolutePath;
            normalizedRequestParameters = NormalizeRequestParameters(parameters);
            StringBuilder signatureBase = new StringBuilder();
            signatureBase.AppendFormat("{0}&", httpMethod.ToUpper());
            signatureBase.AppendFormat("{0}&", UrlEncode(normalizedUrl));
            signatureBase.AppendFormat("{0}", UrlEncode(normalizedRequestParameters));
            return signatureBase.ToString();
        }
        
        public string GenerateSignatureUsingHash(string signatureBase, HashAlgorithm hash)
        {
            return ComputeHash(hash, signatureBase);
        }
       
        public string GenerateSignature(Uri url, string consumerKey, string consumerSecret, string token, string tokenSecret, string httpMethod, string timeStamp, string nonce, out string normalizedUrl, out string normalizedRequestParameters)
        {
            return GenerateSignature(url, consumerKey, consumerSecret, token, tokenSecret, httpMethod, timeStamp, nonce, SignatureTypes.HMACSHA1, out normalizedUrl, out normalizedRequestParameters);
        }
        public string GenerateSignature256(Uri url, string consumerKey, string consumerSecret, string token, string tokenSecret, string httpMethod, string timeStamp, string nonce, out string normalizedUrl, out string normalizedRequestParameters)
        {
            return GenerateSignature(url, consumerKey, consumerSecret, token, tokenSecret, httpMethod, timeStamp, nonce, SignatureTypes.HMACSHA256, out normalizedUrl, out normalizedRequestParameters);
        }
       
        public string GenerateSignature(Uri url, string consumerKey, string consumerSecret, string token, string tokenSecret, string httpMethod, string timeStamp, string nonce, SignatureTypes signatureType, out string normalizedUrl, out string normalizedRequestParameters)
        {
            normalizedUrl = null;
            normalizedRequestParameters = null;
            switch (signatureType)
            {
                case SignatureTypes.PLAINTEXT:
                    //return HttpUtility.UrlEncode(string.Format("{0}&{1}", consumerSecret, tokenSecret));
                    throw new NotImplementedException();
                case SignatureTypes.HMACSHA1:
                    string signatureBase = GenerateSignatureBase(url, consumerKey, token, tokenSecret, httpMethod, timeStamp, nonce, HMACSHA1SignatureType, out normalizedUrl, out normalizedRequestParameters);
                    HMACSHA1 hmacsha1 = new HMACSHA1();
                    
                    hmacsha1.Key = Encoding.ASCII.GetBytes(string.Format("{0}&{1}", UrlEncode(consumerSecret), string.IsNullOrEmpty(tokenSecret) ? "" : UrlEncode(tokenSecret)));
                    return GenerateSignatureUsingHash(signatureBase, hmacsha1);
                case SignatureTypes.HMACSHA256:
                    string signatureBase256 = GenerateSignatureBase(url, consumerKey, token, tokenSecret, httpMethod, timeStamp, nonce, HMACSHA256SignatureType, out normalizedUrl, out normalizedRequestParameters);
                    HMACSHA256 hmacsha256 = new HMACSHA256();
                    
                    hmacsha256.Key = Encoding.ASCII.GetBytes(string.Format("{0}&{1}", UrlEncode(consumerSecret), string.IsNullOrEmpty(tokenSecret) ? "" : UrlEncode(tokenSecret)));
                    return GenerateSignatureUsingHash(signatureBase256, hmacsha256);
                case SignatureTypes.RSASHA1:
                    throw new NotImplementedException();
                default:
                    throw new ArgumentException("Unknown signature type", "signatureType");
            }
        }
      
        public virtual string GenerateTimeStamp()
        {            
            TimeSpan ts = DateTime.UtcNow - new DateTime(1970, 1, 1, 0, 0, 0, 0);
            return Convert.ToInt64(ts.TotalSeconds).ToString();        
            
        }

        public virtual string GenerateNonce(ITracingService tracingService, List<string> errorStack)
        {
            string nonce = EDServicesHelper.getNetSuiteNonce(tracingService, errorStack);
            return nonce;

        }


    }
}

File: Agentic AI Framework - Azure Functions/oauth_base.py
import hashlib
import hmac
import base64
import time
import random
import string
from urllib.parse import urlparse, parse_qs, quote
from typing import Tuple, List, Optional
from dataclasses import dataclass

@dataclass
class QueryParameter:
    """Represents a query parameter for OAuth"""
    name: str
    value: str

    def __lt__(self, other):
        """Comparison for sorting"""
        if self.name == other.name:
            return self.value < other.value
        return self.name < other.name


class SignatureTypes:
    """OAuth signature types enumeration"""
    HMACSHA1 = "HMAC-SHA1"
    HMACSHA256 = "HMAC-SHA256"
    PLAINTEXT = "PLAINTEXT"
    RSASHA1 = "RSA-SHA1"


class OAuthBase:
    """
    OAuth Base class for generating signatures and tokens
    Port of the C# OAuthBase class
    """
    
    # OAuth constants
    OAUTH_VERSION = "1.0"
    OAUTH_PARAMETER_PREFIX = "oauth_"
    OAUTH_CONSUMER_KEY = "oauth_consumer_key"
    OAUTH_CALLBACK = "oauth_callback"
    OAUTH_VERSION_KEY = "oauth_version"
    OAUTH_SIGNATURE_METHOD = "oauth_signature_method"
    OAUTH_SIGNATURE = "oauth_signature"
    OAUTH_TIMESTAMP = "oauth_timestamp"
    OAUTH_NONCE = "oauth_nonce"
    OAUTH_TOKEN = "oauth_token"
    OAUTH_TOKEN_SECRET = "oauth_token_secret"
    
    # Signature type constants
    HMACSHA1_SIGNATURE_TYPE = "HMAC-SHA1"
    HMACSHA256_SIGNATURE_TYPE = "HMAC-SHA256"
    PLAINTEXT_SIGNATURE_TYPE = "PLAINTEXT"
    RSASHA1_SIGNATURE_TYPE = "RSA-SHA1"
    
    # Unreserved characters for URL encoding
    UNRESERVED_CHARS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_.~"
    
    def __init__(self):
        self.random = random.Random()
    
    def url_encode(self, value: str) -> str:
        """
        URL encode a string using OAuth specifications
        
        Args:
            value: String to encode
            
        Returns:
            URL encoded string
        """
        result = []
        for char in value:
            if char in self.UNRESERVED_CHARS:
                result.append(char)
            else:
                result.append('%' + format(ord(char), '02X'))
        return ''.join(result)
    
    def compute_hash(self, algorithm: str, data: str) -> str:
        """
        Compute hash using specified algorithm
        
        Args:
            algorithm: Hash algorithm to use
            data: Data to hash
            
        Returns:
            Base64 encoded hash
        """
        data_bytes = data.encode('ascii')
        
        if algorithm == 'sha1':
            hash_obj = hashlib.sha1(data_bytes)
        elif algorithm == 'sha256':
            hash_obj = hashlib.sha256(data_bytes)
        else:
            raise ValueError(f"Unsupported hash algorithm: {algorithm}")
        
        return base64.b64encode(hash_obj.digest()).decode('ascii')
    
    def get_query_parameters(self, query_string: str) -> List[QueryParameter]:
        """
        Parse query string into list of QueryParameter objects
        
        Args:
            query_string: Query string to parse
            
        Returns:
            List of QueryParameter objects
        """
        if query_string.startswith('?'):
            query_string = query_string[1:]
        
        result = []
        if query_string:
            pairs = query_string.split('&')
            for pair in pairs:
                if pair and not pair.startswith(self.OAUTH_PARAMETER_PREFIX):
                    if '=' in pair:
                        name, value = pair.split('=', 1)
                        result.append(QueryParameter(name, value))
                    else:
                        result.append(QueryParameter(pair, ''))
        
        return result
    
    def normalize_request_parameters(self, parameters: List[QueryParameter]) -> str:
        """
        Normalize request parameters for signature generation
        
        Args:
            parameters: List of QueryParameter objects
            
        Returns:
            Normalized parameter string
        """
        parts = []
        for param in parameters:
            parts.append(f"{param.name}={param.value}")
        return '&'.join(parts)
    
    def generate_signature_base(
        self,
        url: str,
        consumer_key: str,
        token: str,
        token_secret: str,
        http_method: str,
        timestamp: str,
        nonce: str,
        signature_type: str
    ) -> Tuple[str, str, str]:
        """
        Generate signature base string
        
        Args:
            url: Request URL
            consumer_key: OAuth consumer key
            token: OAuth token
            token_secret: OAuth token secret
            http_method: HTTP method (GET, POST, etc.)
            timestamp: OAuth timestamp
            nonce: OAuth nonce
            signature_type: Signature type
            
        Returns:
            Tuple of (signature_base, normalized_url, normalized_parameters)
        """
        token = token or ''
        token_secret = token_secret or ''
        
        if not consumer_key:
            raise ValueError("consumer_key is required")
        if not http_method:
            raise ValueError("http_method is required")
        if not signature_type:
            raise ValueError("signature_type is required")
        
        # Parse URL
        parsed_url = urlparse(url)
        
        # Get query parameters
        parameters = self.get_query_parameters(parsed_url.query)
        
        # Add OAuth parameters
        parameters.append(QueryParameter(self.OAUTH_VERSION_KEY, self.OAUTH_VERSION))
        parameters.append(QueryParameter(self.OAUTH_NONCE, nonce))
        parameters.append(QueryParameter(self.OAUTH_TIMESTAMP, timestamp))
        parameters.append(QueryParameter(self.OAUTH_SIGNATURE_METHOD, signature_type))
        parameters.append(QueryParameter(self.OAUTH_CONSUMER_KEY, consumer_key))
        
        if token:
            parameters.append(QueryParameter(self.OAUTH_TOKEN, token))
        
        # Sort parameters
        parameters.sort()
        
        # Build normalized URL
        normalized_url = f"{parsed_url.scheme}://{parsed_url.hostname}"
        
        # Add port if not default
        if not ((parsed_url.scheme == 'http' and parsed_url.port == 80) or 
                (parsed_url.scheme == 'https' and parsed_url.port == 443)):
            if parsed_url.port:
                normalized_url += f":{parsed_url.port}"
        
        normalized_url += parsed_url.path
        
        # Normalize parameters
        normalized_parameters = self.normalize_request_parameters(parameters)
        
        # Build signature base
        signature_base = (
            f"{http_method.upper()}&"
            f"{self.url_encode(normalized_url)}&"
            f"{self.url_encode(normalized_parameters)}"
        )
        
        return signature_base, normalized_url, normalized_parameters
    
    def generate_signature_using_hash(self, signature_base: str, key: bytes) -> str:
        """
        Generate signature using HMAC
        
        Args:
            signature_base: Signature base string
            key: HMAC key
            
        Returns:
            Base64 encoded signature
        """
        signature_base_bytes = signature_base.encode('ascii')
        hash_obj = hmac.new(key, signature_base_bytes, hashlib.sha256)
        return base64.b64encode(hash_obj.digest()).decode('ascii')
    
    def generate_signature(
        self,
        url: str,
        consumer_key: str,
        consumer_secret: str,
        token: str,
        token_secret: str,
        http_method: str,
        timestamp: str,
        nonce: str,
        signature_type: str = SignatureTypes.HMACSHA1
    ) -> Tuple[str, str, str]:
        """
        Generate OAuth signature
        
        Args:
            url: Request URL
            consumer_key: OAuth consumer key
            consumer_secret: OAuth consumer secret
            token: OAuth token
            token_secret: OAuth token secret
            http_method: HTTP method
            timestamp: OAuth timestamp
            nonce: OAuth nonce
            signature_type: Signature type (default: HMAC-SHA1)
            
        Returns:
            Tuple of (signature, normalized_url, normalized_parameters)
        """
        if signature_type == SignatureTypes.PLAINTEXT:
            raise NotImplementedError("PLAINTEXT signature type not implemented")
        elif signature_type == SignatureTypes.RSASHA1:
            raise NotImplementedError("RSA-SHA1 signature type not implemented")
        elif signature_type == SignatureTypes.HMACSHA1:
            signature_base, normalized_url, normalized_parameters = self.generate_signature_base(
                url, consumer_key, token, token_secret, http_method, timestamp, nonce,
                self.HMACSHA1_SIGNATURE_TYPE
            )
            
            key_string = f"{self.url_encode(consumer_secret)}&{self.url_encode(token_secret) if token_secret else ''}"
            key = key_string.encode('ascii')
            
            signature_base_bytes = signature_base.encode('ascii')
            hash_obj = hmac.new(key, signature_base_bytes, hashlib.sha1)
            signature = base64.b64encode(hash_obj.digest()).decode('ascii')
            
            return signature, normalized_url, normalized_parameters
        elif signature_type == SignatureTypes.HMACSHA256:
            signature_base, normalized_url, normalized_parameters = self.generate_signature_base(
                url, consumer_key, token, token_secret, http_method, timestamp, nonce,
                self.HMACSHA256_SIGNATURE_TYPE
            )
            
            key_string = f"{self.url_encode(consumer_secret)}&{self.url_encode(token_secret) if token_secret else ''}"
            key = key_string.encode('ascii')
            
            signature_base_bytes = signature_base.encode('ascii')
            hash_obj = hmac.new(key, signature_base_bytes, hashlib.sha256)
            signature = base64.b64encode(hash_obj.digest()).decode('ascii')
            
            return signature, normalized_url, normalized_parameters
        else:
            raise ValueError(f"Unknown signature type: {signature_type}")
    
    def generate_signature256(
        self,
        url: str,
        consumer_key: str,
        consumer_secret: str,
        token: str,
        token_secret: str,
        http_method: str,
        timestamp: str,
        nonce: str
    ) -> Tuple[str, str, str]:
        """
        Generate OAuth signature using HMAC-SHA256
        
        Args:
            url: Request URL
            consumer_key: OAuth consumer key
            consumer_secret: OAuth consumer secret
            token: OAuth token
            token_secret: OAuth token secret
            http_method: HTTP method
            timestamp: OAuth timestamp
            nonce: OAuth nonce
            
        Returns:
            Tuple of (signature, normalized_url, normalized_parameters)
        """
        return self.generate_signature(
            url, consumer_key, consumer_secret, token, token_secret,
            http_method, timestamp, nonce, SignatureTypes.HMACSHA256
        )
    
    def generate_timestamp(self) -> str:
        """
        Generate OAuth timestamp
        
        Returns:
            Unix timestamp as string
        """
        return str(int(time.time()))
    
    def generate_nonce(self) -> str:  
        # Generate a random string of 32 characters (fallback or when use_soap_service=False)
        chars = string.ascii_letters + string.digits
        return ''.join(self.random.choice(chars) for _ in range(32))
